core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -120,25 +119,31 @@ private ReadWriteLock getPageLock(PageId pageId) {
   @Override
   public void put(PageId pageId, byte[] page) throws IOException {
     PageId victim = null;
-
+    long victimSize = 0;
+    long pageSize = 0;
     ReadWriteLock pageLock = getPageLock(pageId);
     try (LockResource r = new LockResource(pageLock.writeLock())) {
       boolean alreadyCached;
       boolean needEvict = false;
       try (LockResource r2 = new LockResource(mMetaLock.writeLock())) {
         alreadyCached = mMetaStore.hasPage(pageId);
-        if (!alreadyCached) {
-          needEvict = mPageStore.size() + 1 > mCacheSize;
+        if (alreadyCached) {
+          pageSize = mMetaStore.getPageSize(pageId);
+        } else {
+          needEvict = mPageStore.bytes() + page.length > mCacheSize;
           if (needEvict) {
             victim = mEvictor.evict();
+            victimSize = mMetaStore.getPageSize(victim);
           } else {
-            mMetaStore.addPage(pageId);
+            mMetaStore.addPage(pageId, page.length);
           }
         }
+      } catch (PageNotFoundException e) {
+        throw new IllegalStateException("we shall not reach here");
       }
       if (alreadyCached) {
         try {
-          mPageStore.delete(pageId);
+          mPageStore.delete(pageId, pageSize);
         } catch (PageNotFoundException e) {
           throw new IllegalStateException(
               String.format("Page store is missing page %s.", pageId), e);
```

Comment:
Unrelated question - in some places we do operation first then evictor update and in others we do evictor update then operation, what is the reasoning?

Reply:
that is unintentional. Let's fix that issue in https://github.com/Alluxio/alluxio/pull/10816/

Reply:
fixed in later commit. PTAL